In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("train.csv")
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### Drop the columns which are not required for prediction

In [4]:
data = df.drop(['PassengerId','Name','Ticket','Cabin'],axis=1)
data.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [5]:
data.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [6]:
df = data.dropna()

df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

### Encoding the categorical values of feature

In [7]:
from sklearn.preprocessing import LabelEncoder
from sklearn.compose import ColumnTransformer
label = LabelEncoder()

df['sex_num'] = label.fit_transform(df['Sex'])
df['embarked_num'] = label.fit_transform(df['Embarked'])

df.head()

C:\Users\kashi\AppData\Local\Temp/ipykernel_51968/2079560086.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sex_num'] = label.fit_transform(df['Sex'])
C:\Users\kashi\AppData\Local\Temp/ipykernel_51968/2079560086.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['embarked_num'] = label.fit_transform(df['Embarked'])


Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked  sex_num  \
0         0       3    male  22.0      1      0   7.2500        S        1   
1         1       1  female  38.0      1      0  71.2833        C        0   
2         1       3  female  26.0      0      0   7.9250        S        0   
3         1       1  female  35.0      1      0  53.1000        S        0   
4         0       3    male  35.0      0      0   8.0500        S        1   

   embarked_num  
0             2  
1             0  
2             2  
3             2  
4             2

In [8]:
df.drop(['Sex','Embarked'],axis=1,inplace=True)
df.head()

C:\Users\kashi\Anaconda\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


Survived  Pclass   Age  SibSp  Parch     Fare  sex_num  embarked_num
0         0       3  22.0      1      0   7.2500        1             2
1         1       1  38.0      1      0  71.2833        0             0
2         1       3  26.0      0      0   7.9250        0             2
3         1       1  35.0      1      0  53.1000        0             2
4         0       3  35.0      0      0   8.0500        1             2

### Separating the target feature from other features

In [9]:
X = df.drop(['Survived'],axis=1)
y = df['Survived']

In [10]:
X.head()

Pclass   Age  SibSp  Parch     Fare  sex_num  embarked_num
0       3  22.0      1      0   7.2500        1             2
1       1  38.0      1      0  71.2833        0             0
2       3  26.0      0      0   7.9250        0             2
3       1  35.0      1      0  53.1000        0             2
4       3  35.0      0      0   8.0500        1             2

In [11]:
y.head()

0    0
1    1
2    1
3    1
4    0
Name: Survived, dtype: int64

In [12]:
len(X),len(y)

(712, 712)

### Splitting the data into train and test data

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2)

### Logistic Regression Model

In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix
model1 = LogisticRegression()
model1.fit(X_train,y_train)

C:\Users\kashi\Anaconda\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

### Decision Tree classifier Model

In [15]:
from sklearn.tree import DecisionTreeClassifier
model2 = DecisionTreeClassifier()
model2.fit(X_train,y_train)

DecisionTreeClassifier()

In [16]:
print("Logistic Regression score is: ",model1.score(X_test,y_test))
print("Decision Tree Classifier: ",model2.score(X_test,y_test))

Logistic Regression score is:  0.8041958041958042
Decision Tree Classifier:  0.7552447552447552


In [17]:
predict = model2.predict(X_test)

In [18]:
confusion_matrix(y_test,predict)

array([[69, 21],
       [14, 39]], dtype=int64)

### The Logistic Regression Model shows accuracy of 80.41% while Decision Tree Classifier model shows accuracy of 75.52%

#### The confusion matrix for decision tree classifier shows that : 
- 69 values were predicted correctly to be 'NO' while 14 values were falsely predicted as negative.
- 39 values were predicted correctly to be 'YES' while 29 values falsely predicted positive.


In [20]:
predict = model1.predict(X_test)

In [21]:
confusion_matrix(y_test,predict)

array([[75, 15],
       [13, 40]], dtype=int64)

#### The confusion matrix for logistic regression shows that : 
- 75 values were predicted correctly to be 'NO' while 13 values were falsely predicted as negative.
- 40 values were predicted correctly to be 'YES' while 15 values falsely predicted positive.
